File: easycontrol/server/src/main/java/top/saymzx/easycontrol/server/wrappers/InputManager.java
/*
 * 本项目大量借鉴学习了开源投屏软件：Scrcpy，在此对该项目表示感谢
 */
package top.saymzx.easycontrol.server.wrappers;

import android.view.InputEvent;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public final class InputManager {

  public static final int INJECT_INPUT_EVENT_MODE_ASYNC = 0;

  private static Object manager;
  private static Method injectInputEventMethod;
  private static Method setDisplayIdMethod;

  public static void init(Object m) throws NoSuchMethodException {
    manager = m;
    injectInputEventMethod = manager.getClass().getMethod("injectInputEvent", InputEvent.class, int.class);
    try {
      setDisplayIdMethod = InputEvent.class.getMethod("setDisplayId", int.class);
    } catch (Exception ignored) {
    }
  }

  public static void setDisplayId(InputEvent inputEvent, int displayId) throws InvocationTargetException, IllegalAccessException {
    if (setDisplayIdMethod == null) return;
    setDisplayIdMethod.invoke(inputEvent, displayId);
  }

  public static void injectInputEvent(InputEvent inputEvent, int mode) throws InvocationTargetException, IllegalAccessException {
    injectInputEventMethod.invoke(manager, inputEvent, mode);
  }
}

File: scrcpy_android/server/src/main/java/top/saymzx/scrcpy/server/wrappers/InputManager.kt
package top.saymzx.scrcpy.server.wrappers

import android.view.InputEvent
import android.view.MotionEvent
import java.lang.reflect.Method

class InputManager(private val manager: Any) {
  @get:Throws(NoSuchMethodException::class)
  private var injectInputEventMethod: Method = manager.javaClass.getMethod(
    "injectInputEvent",
    InputEvent::class.java,
    Int::class.javaPrimitiveType
  )

  fun injectInputEvent(inputEvent: InputEvent?, mode: Int): Boolean {
    return try {
      val method = injectInputEventMethod
      method.invoke(manager, inputEvent, mode) as Boolean
    } catch (e: Exception) {
      false
    }
  }

  companion object {
    const val INJECT_INPUT_EVENT_MODE_ASYNC = 0
    const val INJECT_INPUT_EVENT_MODE_WAIT_FOR_RESULT = 1
    const val INJECT_INPUT_EVENT_MODE_WAIT_FOR_FINISH = 2

    @get:Throws(NoSuchMethodException::class)
    private var setDisplayIdMethod: Method? = null
      get() {
        if (field == null) {
          field = InputEvent::class.java.getMethod("setDisplayId", Int::class.javaPrimitiveType)
        }
        return field
      }

    @get:Throws(NoSuchMethodException::class)
    private var setActionButtonMethod: Method? = null
      get() {
        if (field == null) {
          field = MotionEvent::class.java.getMethod("setActionButton", Int::class.javaPrimitiveType)
        }
        return field
      }

    fun setDisplayId(inputEvent: InputEvent?, displayId: Int): Boolean {
      return try {
        val method = setDisplayIdMethod
        method!!.invoke(inputEvent, displayId)
        true
      } catch (e: Exception) {
        false
      }
    }

    fun setActionButton(motionEvent: MotionEvent?, actionButton: Int): Boolean {
      return try {
        val method = setActionButtonMethod
        method!!.invoke(motionEvent, actionButton)
        true
      } catch (e: Exception) {
        false
      }
    }
  }
}